File: src/commonTest/kotlin/com/scoreo/application/RenamePlayerUseCaseTest.kt
package com.scoreo.application

import com.scoreo.domain.DomainError
import com.scoreo.domain.model.Player
import com.scoreo.infrastructure.InMemoryPlayerRepository
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class RenamePlayerUseCaseTest {

    @Test
    fun `renames existing player`() {
        val repo = InMemoryPlayerRepository()
        val player = Player(id = "p1", name = "Alice")
        repo.save(player)
        val useCase = RenamePlayerUseCase(repo)

        useCase("p1", "Alicia")

        val renamed = repo.getAll()[0]
        assertEquals("Alicia", renamed.name)
        assertEquals("p1", renamed.id)
    }

    @Test
    fun `preserves player ID after rename`() {
        val repo = InMemoryPlayerRepository()
        val player = Player(id = "p1", name = "Alice")
        repo.save(player)
        val useCase = RenamePlayerUseCase(repo)

        useCase("p1", "Bob")

        assertEquals("p1", repo.getAll()[0].id)
    }

    @Test
    fun `throws when player not found`() {
        val repo = InMemoryPlayerRepository()
        val useCase = RenamePlayerUseCase(repo)

        assertFailsWith<DomainError.NotFound> {
            useCase("nonexistent", "NewName")
        }
    }

    @Test
    fun `trims whitespace from new name`() {
        val repo = InMemoryPlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = RenamePlayerUseCase(repo)

        useCase("p1", "  Bob  ")

        assertEquals("Bob", repo.getAll()[0].name)
    }

    @Test
    fun `throws when new name is blank`() {
        val repo = InMemoryPlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = RenamePlayerUseCase(repo)

        assertFailsWith<DomainError.Validation> {
            useCase("p1", "   ")
        }
    }

    @Test
    fun `throws when name exceeds 50 characters`() {
        val repo = InMemoryPlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = RenamePlayerUseCase(repo)
        val longName = "A".repeat(51)

        val exception = assertFailsWith<DomainError.Validation> {
            useCase("p1", longName)
        }
        assertTrue(exception.message!!.contains("50"))
    }

    @Test
    fun `accepts name of exactly 50 characters`() {
        val repo = InMemoryPlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = RenamePlayerUseCase(repo)
        val name = "A".repeat(50)

        useCase("p1", name)

        assertEquals(50, repo.getAll()[0].name.length)
    }

    @Test
    fun `allows renaming to same name`() {
        val repo = InMemoryPlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = RenamePlayerUseCase(repo)

        useCase("p1", "Alice")

        assertEquals("Alice", repo.getAll()[0].name)
    }

    @Test
    fun `works with inactive players`() {
        val repo = InMemoryPlayerRepository()
        repo.save(Player("p1", "Alice", active = false))
        val useCase = RenamePlayerUseCase(repo)

        useCase("p1", "Bob")

        val inactive = repo.getAll(includeInactive = true)
        assertEquals("Bob", inactive[0].name)
        assertEquals(false, inactive[0].active)
    }
}

File: apps/scoreo/src/application/renamePlayerUseCase.test.ts
import { describe, expect, it } from 'vitest'
import { NotFoundError, ValidationError } from '../domain/model/errors'
import { InMemoryPlayerRepository } from '../infrastructure/testing/inMemoryPlayerRepository'
import { RenamePlayerUseCase } from './renamePlayerUseCase'

describe('RenamePlayerUseCase', () => {
  it('renames existing player', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new RenamePlayerUseCase(repo)

    useCase.invoke('p1', 'Alicia')

    const renamed = repo.getAll()[0]
    expect(renamed.name).toBe('Alicia')
    expect(renamed.id).toBe('p1')
  })

  it('preserves player ID after rename', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new RenamePlayerUseCase(repo)

    useCase.invoke('p1', 'Bob')

    expect(repo.getAll()[0].id).toBe('p1')
  })

  it('throws when player not found', () => {
    const repo = new InMemoryPlayerRepository()
    const useCase = new RenamePlayerUseCase(repo)

    expect(() => useCase.invoke('nonexistent', 'NewName')).toThrow(NotFoundError)
  })

  it('trims whitespace from new name', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new RenamePlayerUseCase(repo)

    useCase.invoke('p1', '  Bob  ')

    expect(repo.getAll()[0].name).toBe('Bob')
  })

  it('throws when new name is blank', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new RenamePlayerUseCase(repo)

    expect(() => useCase.invoke('p1', '   ')).toThrow(ValidationError)
  })

  it('throws when name exceeds 50 characters', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new RenamePlayerUseCase(repo)
    const longName = 'A'.repeat(51)

    let error: unknown
    try {
      useCase.invoke('p1', longName)
    } catch (e) {
      error = e
    }
    expect(error).toBeInstanceOf(ValidationError)
    expect((error as ValidationError).message).toContain('50')
  })

  it('accepts name of exactly 50 characters', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new RenamePlayerUseCase(repo)
    const name = 'A'.repeat(50)

    useCase.invoke('p1', name)

    expect(repo.getAll()[0].name).toHaveLength(50)
  })

  it('allows renaming to same name', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new RenamePlayerUseCase(repo)

    useCase.invoke('p1', 'Alice')

    expect(repo.getAll()[0].name).toBe('Alice')
  })

  it('works with inactive players', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: false })
    const useCase = new RenamePlayerUseCase(repo)

    useCase.invoke('p1', 'Bob')

    const inactive = repo.getAll(true)
    expect(inactive[0].name).toBe('Bob')
    expect(inactive[0].active).toBe(false)
  })
})
